pkg/virt-handler/heartbeat/multipath-socket.go

```diff
@@ -0,0 +1,169 @@
+package heartbeat
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/safepath"
+	"kubevirt.io/kubevirt/pkg/storage/reservation"
+	"kubevirt.io/kubevirt/pkg/virt-handler/heartbeat/filewatcher"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+	virt_chroot "kubevirt.io/kubevirt/pkg/virt-handler/virt-chroot"
+)
+
+const (
+	procRootRun = "/proc/1/root/run"
+	socket      = "multipathd.socket"
+)
+
+//go:generate mockgen -source $GOFILE -package=$GOPACKAGE -destination=generated_mock_$GOFILE
+type mounter interface {
+	Mount(sourcePath, targetPath *safepath.Path, ro bool) *exec.Cmd
+	Umount(path *safepath.Path) *exec.Cmd
+	IsMounted(mountPoint *safepath.Path) (isMounted bool, err error)
+}
+
+type mountManager struct{}
+
+func (m *mountManager) Mount(sourcePath, targetPath *safepath.Path, ro bool) *exec.Cmd {
+	return virt_chroot.MountChroot(sourcePath, targetPath, ro)
+}
+
+func (m *mountManager) Umount(path *safepath.Path) *exec.Cmd {
+	return virt_chroot.UmountChroot(path)
+}
+
+func (m *mountManager) IsMounted(mountPoint *safepath.Path) (isMounted bool, err error) {
+	return isolation.IsMounted(mountPoint)
+}
+
+type MonitorMultipathSocket struct {
+	Watcher             *filewatcher.FileWatcher
+	MultipathSocketPath string
+	HostDir             string
+	Mounter             mounter
+}
+
+func NewMonitorMultipathSocket() *MonitorMultipathSocket {
+	multipathSocketPath := filepath.Join(procRootRun, socket)
+	return &MonitorMultipathSocket{
+		Watcher:             filewatcher.New(multipathSocketPath, 1*time.Second),
+		MultipathSocketPath: multipathSocketPath,
+		HostDir:             reservation.GetPrHelperHostSocketDir(),
+		Mounter:             &mountManager{},
+	}
+}
+
+func (m *MonitorMultipathSocket) Run() {
+	sPath, err := m.createSPath()
+	if err != nil {
+		return
+	}
+
+	// Remove old bind mount if there was one
+	if err := m.unmount(sPath); err != nil {
+		return
+	}
+	// Try to mount existing socket and ignore errors
+	_ = m.mount(sPath)
+
+	m.Watcher.Run()
+	for {
+		select {
+		case event := <-m.Watcher.Events:
+			switch event {
+			case filewatcher.Create:
+				if err := m.mount(sPath); err != nil {
+					continue
+				}
+			case filewatcher.Remove:
+				if err := m.unmount(sPath); err != nil {
+					continue
+				}
+			case filewatcher.InoChange:
+				if err := m.unmount(sPath); err != nil {
+					continue
+				}
+				if err := m.mount(sPath); err != nil {
+					continue
+				}
+			}
+		case err := <-m.Watcher.Errors:
+			if err != nil {
+				log.Log.Reason(err).Errorf("Failed monitoring multipath socket")
+				m.Watcher.Close()
+				return
+			}
+		}
+	}
+}
+
+func (m *MonitorMultipathSocket) Stop() {
```

Comment:
don't we need to clean up here? defer m.mounter.Umount(m.mountPath)?

Reply:
Added clean up to `Run`

Reply:
Thanks!